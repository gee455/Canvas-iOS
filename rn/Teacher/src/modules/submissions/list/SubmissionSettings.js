//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

import React, { PureComponent } from 'react'
import {
  ScrollView,
  StyleSheet,
} from 'react-native'
import Screen from '../../../routing/Screen'
import i18n from 'format-message'
import RowWithSwitch from '../../../common/components/rows/RowWithSwitch'
import RowSeparator from '../../../common/components/rows/RowSeparator'
import { connect } from 'react-redux'
import AssignmentActions from '../../assignments/actions'
import branding from '../../../common/branding'
import { SubTitle } from '../../../common/text'

type SubmissionSettingsOwnProps = {
  courseID: string,
  assignmentID: string,
  navigator: Navigator,
}
type SubmissionSettingsDataProps = {
  anonymous: boolean,
  muted: boolean,
  assignment: Assignment,
  disableAnonymous: boolean,
}
type SubmissionSettingsActions = {
  anonymousGrading: (string, string, boolean) => void,
  updateAssignment: (string, Assignment, Assignment) => void,
}

type SubmissionSettingsProps =
  SubmissionSettingsOwnProps &
  SubmissionSettingsDataProps &
  SubmissionSettingsActions

export class SubmissionSettings extends PureComponent<SubmissionSettingsProps> {
  props: SubmissionSettingsProps

  toggleAnonymousGrading = (value: boolean) => {
    this.props.anonymousGrading(
      this.props.courseID,
      this.props.assignmentID,
      value
    )
  }

  toggleMutedGrading = (value: boolean) => {
    this.props.updateAssignment(
      this.props.courseID,
      {
        ...this.props.assignment,
        muted: value,
      },
      this.props.assignment
    )
  }

  render () {
    return (
      <Screen
        title={i18n('Submission Settings')}
        navBarButtonColor={branding.primaryButtonColor}
      >
        <ScrollView style={style.container}>
          <RowSeparator />
          <RowWithSwitch
            title={i18n('Mute Grades')}
            value={this.props.muted}
            onValueChange={this.toggleMutedGrading}
            identifier='submission-settings.muted'
          />
          <RowSeparator />
          <RowWithSwitch
            title={i18n('Anonymous Grading')}
            value={this.props.anonymous}
            onValueChange={this.toggleAnonymousGrading}
            identifier='submission-settings.anonymous'
            disabled={this.props.disableAnonymous}
          />
          <RowSeparator />
          <SubTitle style={{ paddingHorizontal: 12, paddingVertical: 4 }}>
            {i18n('This will anonymize each student and shuffle the submission list.')}
          </SubTitle>
        </ScrollView>
      </Screen>
    )
  }
}

export function mapStateToProps (state: AppState, ownProps: SubmissionSettingsOwnProps) {
  let course = state.entities.courses[ownProps.courseID]
  let anonymousCourse = course && course.enabledFeatures.includes('anonymous_grading')
  let assignment = state.entities.assignments[ownProps.assignmentID].data
  let muted = assignment.muted
  let quiz = assignment.quiz_id && state.entities.quizzes[assignment.quiz_id]
  let anonymousQuiz = quiz && quiz.data && quiz.data.anonymous_submissions

  let anonymous = state.entities.assignments[ownProps.assignmentID].anonymousGradingOn || anonymousCourse || anonymousQuiz
  let disableAnonymous = anonymousCourse || anonymousQuiz || false

  return { anonymous, muted, assignment, disableAnonymous }
}
const Connect = connect(mapStateToProps, AssignmentActions)(SubmissionSettings)
export default (Connect: any)

const style = StyleSheet.create({
  container: {
    flex: 1,
  },
})
