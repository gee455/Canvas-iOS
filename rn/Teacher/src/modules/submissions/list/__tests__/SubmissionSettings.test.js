//
// Copyright (C) 2016-present Instructure, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

/* eslint-disable flowtype/require-valid-file-annotation */

import React from 'react'
import { SubmissionSettings, mapStateToProps } from '../SubmissionSettings'
import renderer from 'react-test-renderer'
import explore from '../../../../../test/helpers/explore'

const template = {
  ...require('../../../../__templates__/helm'),
  ...require('../../../../__templates__/assignments'),
  ...require('../../../../__templates__/quiz'),
  ...require('../../../../redux/__templates__/app-state'),
}

let defaultProps = {
  courseID: '1',
  assignmentID: '2',
  navigator: template.navigator(),
  anonymousGrading: jest.fn(),
  updateAssignment: jest.fn(),
  anonymous: false,
  muted: false,
  assignment: template.assignment({ id: '2' }),
}

describe('SubmissionSettings', () => {
  beforeEach(() => jest.resetAllMocks())

  it('renders properly', () => {
    let tree = renderer.create(
      <SubmissionSettings {...defaultProps} />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders the anonymous grading toggle as disabled when disableAnonymous is true', () => {
    let tree = renderer.create(
      <SubmissionSettings {...defaultProps} disableAnonymous={true} />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('calls anonymousGrading when the toggle is pressed', () => {
    let tree = renderer.create(
      <SubmissionSettings {...defaultProps} />
    )

    let toggle = explore(tree.toJSON()).selectByID('submission-settings.anonymous') || {}
    toggle.props.onValueChange(true)

    expect(defaultProps.anonymousGrading).toHaveBeenCalledWith(
      '1', '2', true
    )
  })

  it('calls updateAssignment when mute toggle is pressed', () => {
    let tree = renderer.create(
      <SubmissionSettings {...defaultProps} />
    )

    let toggle = explore(tree.toJSON()).selectByID('submission-settings.muted') || {}
    toggle.props.onValueChange(true)

    expect(defaultProps.updateAssignment).toHaveBeenCalledWith(
      '1',
      { ...defaultProps.assignment, muted: true },
      defaultProps.assignment
    )
  })
})

describe('mapStateToProps', () => {
  let ownProps = {
    courseID: '1',
    assignmentID: '1',
  }

  it('disables anonymous and anonymous is true when the course is anonymous', () => {
    let state = template.appState({
      entities: {
        courses: {
          '1': {
            enabledFeatures: ['anonymous_grading'],
          },
        },
        assignments: {
          '1': {
            data: template.assignment(),
          },
        },
      },
    })

    let props = mapStateToProps(state, ownProps)
    expect(props).toMatchObject({
      anonymous: true,
      disableAnonymous: true,
    })
  })

  it('disables anonymous and anonymous is truen when the assignment is associated to a quiz and it has anonymous_submissions turned on', () => {
    let state = template.appState({
      entities: {
        courses: {
          '1': {
            enabledFeatures: [],
          },
        },
        assignments: {
          '1': {
            data: template.assignment({ id: '1', quiz_id: '1' }),
            anonymousGradingOn: false,
          },
        },
        quizzes: {
          '1': {
            data: template.quiz({ id: '1', anonymous_submissions: true }),
          },
        },
      },
    })

    let props = mapStateToProps(state, ownProps)
    expect(props).toMatchObject({
      anonymous: true,
      disableAnonymous: true,
    })
  })

  it('doesnt disable anonymous and anonymous is true when the assignment has anonymous grading turned on', () => {
    let state = template.appState({
      entities: {
        courses: {
          '1': {
            enabledFeatures: [],
          },
        },
        assignments: {
          '1': {
            data: template.assignment(),
            anonymousGradingOn: true,
          },
        },
      },
    })

    let props = mapStateToProps(state, ownProps)
    expect(props).toMatchObject({
      anonymous: true,
      disableAnonymous: false,
    })
  })

  it('returns the muted value of the assignment', () => {
    let state = template.appState({
      entities: {
        courses: {
          '1': {
            enabledFeatures: [],
          },
        },
        assignments: {
          '1': {
            data: template.assignment({ muted: true }),
            anonymousGradingOn: false,
          },
        },
      },
    })

    let props = mapStateToProps(state, ownProps)
    expect(props.muted).toEqual(true)
  })

  it('returns the assignment', () => {
    let state = template.appState({
      entities: {
        courses: {
          '1': {
            enabledFeatures: [],
          },
        },
        assignments: {
          '1': {
            data: template.assignment(),
            anonymousGradingOn: false,
          },
        },
      },
    })

    let props = mapStateToProps(state, ownProps)
    expect(props.assignment).toEqual(state.entities.assignments['1'].data)
  })
})
