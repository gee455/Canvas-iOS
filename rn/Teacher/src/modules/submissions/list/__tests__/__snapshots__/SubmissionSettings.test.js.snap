// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmissionSettings renders properly 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#C7CDD1",
              "height": 0.5,
            },
            undefined,
          ]
        }
      />
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel="Mute Grades, Off"
        accessibilityTraits={
          Array [
            "button",
          ]
        }
        accessible={true}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            Array [
              undefined,
              undefined,
            ],
          ]
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "flex": 1,
                "flexDirection": "row",
                "minHeight": 54,
                "paddingBottom": 12,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 12,
              },
              Object {
                "backgroundColor": "white",
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "center",
                },
                Object {
                  "marginLeft": 0,
                },
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              numberOfLines={0}
              style={
                Array [
                  Object {
                    "fontFamily": ".SFUIDisplay",
                  },
                  Object {
                    "color": "#2D3B45",
                    "fontSize": 16,
                  },
                  Array [
                    Object {
                      "fontWeight": "600",
                    },
                  ],
                  Object {
                    "fontFamily": ".SFUIDisplay-semibold",
                  },
                ]
              }
            >
              Mute Grades
            </Text>
          </View>
          <View
            style={
              Object {
                "flex": 0,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "flex-end",
                }
              }
            >
              <RCTSwitch
                disabled={false}
                identifier="submission-settings.muted"
                onChange={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                onTintColor="#374A59"
                onValueChange={[Function]}
                style={
                  Array [
                    Object {
                      "height": 31,
                      "width": 51,
                    },
                    undefined,
                  ]
                }
                testID="submission-settings.muted"
                tintColor={null}
                title="Mute Grades"
                value={false}
              />
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#C7CDD1",
              "height": 0.5,
            },
            undefined,
          ]
        }
      />
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel="Anonymous Grading, Off"
        accessibilityTraits={
          Array [
            "button",
          ]
        }
        accessible={true}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            Array [
              undefined,
              undefined,
            ],
          ]
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "flex": 1,
                "flexDirection": "row",
                "minHeight": 54,
                "paddingBottom": 12,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 12,
              },
              Object {
                "backgroundColor": "white",
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "center",
                },
                Object {
                  "marginLeft": 0,
                },
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              numberOfLines={0}
              style={
                Array [
                  Object {
                    "fontFamily": ".SFUIDisplay",
                  },
                  Object {
                    "color": "#2D3B45",
                    "fontSize": 16,
                  },
                  Array [
                    Object {
                      "fontWeight": "600",
                    },
                  ],
                  Object {
                    "fontFamily": ".SFUIDisplay-semibold",
                  },
                ]
              }
            >
              Anonymous Grading
            </Text>
          </View>
          <View
            style={
              Object {
                "flex": 0,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "flex-end",
                }
              }
            >
              <RCTSwitch
                disabled={false}
                identifier="submission-settings.anonymous"
                onChange={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                onTintColor="#374A59"
                onValueChange={[Function]}
                style={
                  Array [
                    Object {
                      "height": 31,
                      "width": 51,
                    },
                    undefined,
                  ]
                }
                testID="submission-settings.anonymous"
                tintColor={null}
                title="Anonymous Grading"
                value={false}
              />
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#C7CDD1",
              "height": 0.5,
            },
            undefined,
          ]
        }
      />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "fontFamily": ".SFUIDisplay",
            },
            Object {
              "color": "#7F91A7",
              "fontSize": 14,
            },
            Object {
              "paddingHorizontal": 12,
              "paddingVertical": 4,
            },
          ]
        }
      >
        This will anonymize each student and shuffle the submission list.
      </Text>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;

exports[`SubmissionSettings renders the anonymous grading toggle as disabled when disableAnonymous is true 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#C7CDD1",
              "height": 0.5,
            },
            undefined,
          ]
        }
      />
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel="Mute Grades, Off"
        accessibilityTraits={
          Array [
            "button",
          ]
        }
        accessible={true}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            Array [
              undefined,
              undefined,
            ],
          ]
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "flex": 1,
                "flexDirection": "row",
                "minHeight": 54,
                "paddingBottom": 12,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 12,
              },
              Object {
                "backgroundColor": "white",
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "center",
                },
                Object {
                  "marginLeft": 0,
                },
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              numberOfLines={0}
              style={
                Array [
                  Object {
                    "fontFamily": ".SFUIDisplay",
                  },
                  Object {
                    "color": "#2D3B45",
                    "fontSize": 16,
                  },
                  Array [
                    Object {
                      "fontWeight": "600",
                    },
                  ],
                  Object {
                    "fontFamily": ".SFUIDisplay-semibold",
                  },
                ]
              }
            >
              Mute Grades
            </Text>
          </View>
          <View
            style={
              Object {
                "flex": 0,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "flex-end",
                }
              }
            >
              <RCTSwitch
                disabled={false}
                identifier="submission-settings.muted"
                onChange={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                onTintColor="#374A59"
                onValueChange={[Function]}
                style={
                  Array [
                    Object {
                      "height": 31,
                      "width": 51,
                    },
                    undefined,
                  ]
                }
                testID="submission-settings.muted"
                tintColor={null}
                title="Mute Grades"
                value={false}
              />
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#C7CDD1",
              "height": 0.5,
            },
            undefined,
          ]
        }
      />
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel="Anonymous Grading, Off"
        accessibilityTraits={
          Array [
            "disabled",
            "button",
          ]
        }
        accessible={true}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            Array [
              undefined,
              undefined,
            ],
          ]
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "flex": 1,
                "flexDirection": "row",
                "minHeight": 54,
                "paddingBottom": 12,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 12,
              },
              Object {
                "backgroundColor": "white",
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "center",
                },
                Object {
                  "marginLeft": 0,
                },
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              numberOfLines={0}
              style={
                Array [
                  Object {
                    "fontFamily": ".SFUIDisplay",
                  },
                  Object {
                    "color": "#2D3B45",
                    "fontSize": 16,
                  },
                  Array [
                    Object {
                      "fontWeight": "600",
                    },
                  ],
                  Object {
                    "fontFamily": ".SFUIDisplay-semibold",
                  },
                ]
              }
            >
              Anonymous Grading
            </Text>
          </View>
          <View
            style={
              Object {
                "flex": 0,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "flex-end",
                }
              }
            >
              <RCTSwitch
                disabled={true}
                identifier="submission-settings.anonymous"
                onChange={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                onTintColor="#374A59"
                onValueChange={[Function]}
                style={
                  Array [
                    Object {
                      "height": 31,
                      "width": 51,
                    },
                    undefined,
                  ]
                }
                testID="submission-settings.anonymous"
                tintColor={null}
                title="Anonymous Grading"
                value={false}
              />
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#C7CDD1",
              "height": 0.5,
            },
            undefined,
          ]
        }
      />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "fontFamily": ".SFUIDisplay",
            },
            Object {
              "color": "#7F91A7",
              "fontSize": 14,
            },
            Object {
              "paddingHorizontal": 12,
              "paddingVertical": 4,
            },
          ]
        }
      >
        This will anonymize each student and shuffle the submission list.
      </Text>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
